src/util/AbstractMemory/Memory.cpp

```diff
@@ -0,0 +1,116 @@
+double Memory::kilobytes() const {
+  return sizetDivision(memoryInBytes_, numberOfBytesPerKB);
+}
+
+// _____________________________________________________________________________
+double Memory::megabytes() const {
+  return sizetDivision(memoryInBytes_, numberOfBytesPerMB);
+}
+
+// _____________________________________________________________________________
+double Memory::gigabytes() const {
+  return sizetDivision(memoryInBytes_, numberOfBytesPerGB);
+}
+
+// _____________________________________________________________________________
+double Memory::terabytes() const {
+  return sizetDivision(memoryInBytes_, numberOfBytesPerTB);
+}
+
+// _____________________________________________________________________________
+double Memory::petabytes() const {
+  return sizetDivision(memoryInBytes_, numberOfBytesPerPB);
+}
+
+// _____________________________________________________________________________
+std::string Memory::asString() const {
+  // Convert number and memory unit name to the string, we want to return.
+  auto toString = [](const auto& number, std::string_view unitName) {
+    return absl::StrCat(number, " ", unitName);
+  };
+
+  // Just use the first unit, which is bigger/equal to `memoryInBytes_`.
+  if (memoryInBytes_ >= numberOfBytesPerPB) {
+    return toString(petabytes(), "PB");
+  } else if (memoryInBytes_ >= numberOfBytesPerTB) {
+    return toString(terabytes(), "TB");
+  } else if (memoryInBytes_ >= numberOfBytesPerGB) {
+    return toString(gigabytes(), "GB");
+  } else if (memoryInBytes_ >= numberOfBytesPerMB) {
+    return toString(megabytes(), "MB");
+  } else if (memoryInBytes_ >= numberOfBytesPerKB) {
+    return toString(kilobytes(), "KB");
+  } else {
+    // Just return the amount of bytes.
+    return toString(memoryInBytes_, "Byte");
+  }
+}
+
+// _____________________________________________________________________________
+void Memory::parse(std::string_view str) {
+  antlr4::ANTLRInputStream input(str);
+  MemoryDefinitionLanguageLexer lexer(&input);
+  antlr4::CommonTokenStream tokens(&lexer);
+  MemoryDefinitionLanguageParser parser(&tokens);
+
+  // The default in ANTLR is to log all errors to the console and to continue
+  // the parsing. We need to turn parse errors into exceptions instead to
+  // propagate them to the user.
+  ThrowingErrorListener errorListener{};
+  parser.removeErrorListeners();
+  parser.addErrorListener(&errorListener);
+  lexer.removeErrorListeners();
+  lexer.addErrorListener(&errorListener);
+
+  // Get the top node. That is, the node of the first grammar rule.
+  MemoryDefinitionLanguageParser::MemoryDefinitionStringContext*
+      memoryDefinitionStringContext{parser.memoryDefinitionString()};
+
+  // Our visitor should now convert this to an amount of bytes, we can simply
+  // assign.
+  memoryInBytes_ =
+      ToSizeTMemoryDefinitionLanguageVisitor{}.visitMemoryDefinitionString(
+          memoryDefinitionStringContext);
```

Comment:
Do we get sufficiently good error messages when parsing?
Most notably, the parse error should give some examples for valid memory sizes.

Reply:
No, we only got standard parser error messages. I got no experience with writing error listeners for the parser, so this will take a while.

Reply:
Well I don't think if the error listeners are worth it here.
I would just catch the internal errors and write something hardcoded like
` <your input> could not be parsed as a memory size. Examples for valid memory sizes are "4 B", "3.21 MB", "2.392 TB".

Reply:
Alright, how does that look: [72c9b93](https://github.com/ad-freiburg/qlever/pull/1032/commits/72c9b93db4cc17be76a9343c046a8f503aebb396)